libexec/apply_catalog.rb

```diff
@@ -13,34 +13,22 @@
 
 # Create temporary directories for all core Puppet settings so we don't clobber
 # existing state or read from puppet.conf
+Puppet.initialize_settings
 puppet_root = Dir.mktmpdir
+Puppet::Settings::REQUIRED_APP_SETTINGS.each do |setting|
+  Puppet[setting] = File.join(puppet_root, setting.to_s.chomp('dir'))
+end
```

Comment:
I think we should explicitly call puppet.settings.use( ??? ) to make it easier to understand and avoid regressions.

Not sure exactly where that should happen. before line 25 maybe?